block/provider/delete_snapshot.go

```diff
@@ -18,31 +18,65 @@
 package provider
 
 import (
+	"time"
+
+	"github.com/IBM/ibmcloud-volume-interface/lib/metrics"
 	"github.com/IBM/ibmcloud-volume-interface/lib/provider"
 	userError "github.com/IBM/ibmcloud-volume-vpc/common/messages"
+	"github.com/IBM/ibmcloud-volume-vpc/common/vpcclient/models"
 	"go.uber.org/zap"
 )
 
 // DeleteSnapshot delete snapshot
 func (vpcs *VPCSession) DeleteSnapshot(snapshot *provider.Snapshot) error {
-	vpcs.Logger.Info("Entry DeleteSnapshot", zap.Reflect("snapshot", snapshot))
-	defer vpcs.Logger.Info("Exit DeleteSnapshot", zap.Reflect("snapshot", snapshot))
+	vpcs.Logger.Info("Entry DeleteSnapshot", zap.Reflect("snapshotID", snapshot.SnapshotID))
+	defer vpcs.Logger.Info("Exit DeleteSnapshot", zap.Reflect("snapshotID", snapshot.SnapshotID))
+	defer metrics.UpdateDurationFromStart(vpcs.Logger, "DeleteSnapshot", time.Now())
 
 	var err error
-	_, err = vpcs.GetSnapshot(snapshot.SnapshotID)
-	if err != nil {
-		return userError.GetUserError("StorageFindFailedWithSnapshotId", err, snapshot.SnapshotID, "Not a valid snapshot ID")
+	if snapshot == nil {
+		err = userError.GetUserError("InvalidSnapshotID", nil, nil)
+		return err
 	}
 
+	vpcs.Logger.Info("Deleting snapshot from VPC provider...")
 	err = retry(vpcs.Logger, func() error {
-		err = vpcs.Apiclient.SnapshotService().DeleteSnapshot(snapshot.Volume.VolumeID, snapshot.SnapshotID, vpcs.Logger)
+		err = vpcs.Apiclient.SnapshotService().DeleteSnapshot(snapshot.SnapshotID, vpcs.Logger)
 		return err
 	})
 
 	if err != nil {
 		return userError.GetUserError("FailedToDeleteSnapshot", err)
 	}
 
+	err = WaitForSnapshotDeletion(vpcs, snapshot.SnapshotID)
+	if err != nil {
+		return userError.GetUserError("FailedToDeleteSnapshot", err, snapshot.SnapshotID)
+	}
 	vpcs.Logger.Info("Successfully deleted the snapshot with backend (vpcclient) call)")
-	return nil
+	return err
+}
+
+// WaitForSnapshotDeletion checks the snapshot for valid status
+func WaitForSnapshotDeletion(vpcs *VPCSession, snapshotID string) (err error) {
+	vpcs.Logger.Debug("Entry of WaitForSnapshotDeletion method...")
+	defer vpcs.Logger.Debug("Exit from WaitForSnapshotDeletion method...")
+	var skip = false
+
+	vpcs.Logger.Info("Getting snapshot details from VPC provider...", zap.Reflect("snapshotID", snapshotID))
+
+	err = vpcs.APIRetry.FlexyRetry(vpcs.Logger, func() (error, bool) {
+		_, err = vpcs.Apiclient.SnapshotService().GetSnapshot(snapshotID, vpcs.Logger)
+		// Keep retry, until GetSnapshot returns snapshots_not_found
```

Comment:
check number of retries

Reply:
Driver side Deletion for 450G- test

Reply:
We need to revalidate the exponential interval in performance test. As of now it looks good for me 10 retries with exponential retry starting from 10 seconds and max gap is 60 seconds.

Reply:
https://github.ibm.com/alchemy-containers/armada-storage/issues/3801